pkg/pmem-csi-operator/controller/deployment/deployment_controller.go

```diff
@@ -47,25 +53,128 @@ func Add(mgr manager.Manager, opts pmemcontroller.ControllerOptions) error {
+			if new.DeletionTimestamp != nil {
+				// Deployment CR deleted, remove it's reference from cache.
+				// Objects owned by it are automatically garbage collected.
+				r.deleteDeployment(e.MetaOld.GetName())
+				return false
+			}
+			if old.GetGeneration() == new.GetGeneration() {
+				// No changes registered
+				return false
+			}
+			// We are intersted in only spec changes, not CR status changes
+			changes := old.Compare(new)
+			klog.Infof("CR changes: %v", changes)
+			return len(changes) != 0
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			r.reconcileMutex.Lock()
+			defer r.reconcileMutex.Unlock()
+			klog.Infof("DELETED: %T/%s", e.Object, e.Meta.GetName())
+			// Deployment CR deleted, remove it's reference from cache.
+			// Objects owned by it are automatically garbage collected.
+			r.deleteDeployment(e.Meta.GetName())
+			// We already handled the event here,
+			// so no more further reconcile required.
+			return false
+		},
+	}
+
 	// Watch for changes to primary resource Deployment
-	err = c.Watch(&source.Kind{Type: &pmemcsiv1alpha1.Deployment{}}, &handler.EnqueueRequestForObject{})
-	if err != nil {
+	if err := c.Watch(&source.Kind{Type: &pmemcsiv1alpha1.Deployment{}}, &handler.EnqueueRequestForObject{}, p); err != nil {
 		klog.Errorf("Deployment.Add: watch error: %v", err)
 		return err
 	}
 
+	// Predicate functions for sub-object changes
+	// We do not want to pollute deployment reconcile loop by sending
+	// sub-object changes, instead we provide a dedicated handler.
+	// So all these event handlers returns 'false' so that the event
+	// is not propagated further.
+	// One exception is: If we fail to handle here, then we pass this
+	// event to reconcile loop, where it should recognize these requests
+	// and just requeue. Expecting that the failure is retried.
+	eventFunc := func(meta metav1.Object, obj apiruntime.Object) bool {
+		// Get the owned deployment
+		d, err := r.getDeploymentFor(meta)
+		if err != nil {
+			klog.Infof("%T/%s is not owned by any deployment", obj, meta.GetName())
+			// The owner might have deleted already
+			// we can safely ignore this event
+			return false
+		}
+		r.reconcileMutex.Lock()
+		defer r.reconcileMutex.Unlock()
+		if err := d.HandleEvent(meta, obj, r); err != nil {
+			klog.Warningf("Error occurred while handling the event: %v, requeuing the event", err)
+			return true
+		}
+		return false
+	}
+	sop := predicate.Funcs{
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			klog.Infof("DELETED: %T/%s", e.Object, e.Meta.GetName())
+			return eventFunc(e.Meta, e.Object)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if e.MetaNew.GetDeletionTimestamp() != nil {
+				// We can handle this in delete handler
+				return false
+			}
+			klog.Infof("UPDATED: %T/%s", e.ObjectOld, e.MetaOld.GetName())
+			return eventFunc(e.MetaOld, e.ObjectOld)
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			// Do not handle sub-object create events
+			return false
```

Comment:
There is nothing we should do on object creation. obviously, those events are supposed to be the result of a previous reconcile loop.

Reply:
Theoretically someone might create objects that are owned by the CR which are not needed. We do remove those on operator startup, so it is worth calling out here *in the comment* that we intentionally don't do that because a) it's a rather obscure situation and b) we want to avoid doing work for the usual case where the object was created by us and thus don't need changes.

Reply:
Added the comment, can you please check if that is sufficient.

Reply:
You keep skipping the space before an opening bracket, but let's ignore that for now.

I'm not sure about the "that anyway gets removed in the next reconcile loop" argument - what if there is no such loop iteration for quite a while? The whole point of this callback function is to trigger such a reconciliation as soon as the extra object gets created.

The argument needs to be more along the lines of "this shouldn't happen in practice and checking whether it has happened is just going to cause extra overhead in the normal case of objects that were created by the operator".

Reply:
Okay, rephrased the comment by including the suggested text.